### PR TITLE
feat(olm): support depending on APIservices

### PR DESCRIPTION
**Pull request by ecordell**
- Adds the ability to depend upon apiservices in a cluster
- Doesn't yet support installing Apiservices

### CODE REVIEW

File: deploy/chart/templates/03-clusterserviceversion.crd.yaml [hunk 1]
```diff
@@ -180,6 +180,190 @@ spec:
...
+                            x-descriptors:
+                              type: array
+                              description: A list of descriptors for the status entry that indicate the meaning of the field.
+                              items:
+                                type: string
+                            value:
+                              type: object
+                              description: If present, the value of this status is the same for all instances of the CRD and can be found here instead of on the CR.
+
             customresourcedefinitions:
```
Comment (alecmerdler) [inline deploy/chart/templates/03-clusterserviceversion.crd.yaml]: You just duplicated the `customresourcedefinitions` block, right? The plan is to get rid of it in favor of generic `apiservicedefinitions`?

File: deploy/chart/templates/03-clusterserviceversion.crd.yaml [hunk 2]
```diff
@@ -180,6 +180,190 @@ spec:
...
+                    required:
+                    - name
+                    - version
+                    - kind
+                    - displayName
+                    - description
+                    properties:
+                      name:
+                        type: string
+                        description: Fully qualified name of the CustomResourceDefinition (e.g. my-resource-v1.app.coreos.com)
```
Comment (alecmerdler) [inline deploy/chart/templates/03-clusterserviceversion.crd.yaml]: "Fully qualified name of the APIService"

File: deploy/chart/templates/03-clusterserviceversion.crd.yaml (hunk 1 above)
Comment (ecordell) [reply deploy/chart/templates/03-clusterserviceversion.crd.yaml]: I duplicated it for v1alpha1, and for v2alpha1 the plan is to do something like that ("providedAPIs", "requiredAPIs"). This keeps things simple for backwards-compatibility.

File: pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go [hunk 3]
```diff
@@ -282,3 +303,32 @@ func (csv ClusterServiceVersion) GetAllCRDDescriptions() []CRDDescription {
 
 	return descs
 }
+
+// GetAllAPIServiceDescriptions returns a deduplicated set of CRDDescriptions that is
```
Comment (njhale) [inline pkg/api/apis/operators/v1alpha1/clusterserviceversion_types.go]: nit: CRDDescriptions -> APIServiceDescriptions

**Comment by ecordell** [comment]
I rewrote a chunk of this because I realized I wasn't checking for the availability of the Kind, so I added checking for that via discovery.

File: test/e2e/csv_e2e_test.go [hunk 4]
```diff
@@ -316,6 +372,69 @@ func TestCreateCSVRequirementsMet(t *testing.T) {
...
+	csv := v1alpha1.ClusterServiceVersion{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       v1alpha1.ClusterServiceVersionKind,
+			APIVersion: v1alpha1.ClusterServiceVersionAPIVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: genName("csv"),
+		},
+		Spec: v1alpha1.ClusterServiceVersionSpec{
+			InstallStrategy: v1alpha1.NamedInstallStrategy{
```
Comment (njhale) [inline test/e2e/csv_e2e_test.go]: You could also use [this helper](https://github.com/operator-framework/operator-lifecycle-manager/blob/678fea44cfea340f798b8aa629aa8f388a32b66f/test/e2e/installplan_e2e_test.go#L86).

**Pull request merged**